Add support for smile encoding

nezihyigitbasi opened the pull request.

Commit:
Add support for smile encoding

nezihyigitbasi commented:
@electrum can you please take a look when you have time?

electrum commented on json/src/main/java/io/airlift/json/SmileCodecFactory.java:
```diff
@@ -0,0 +1,155 @@
+    }
+
+    @Inject
+    public SmileCodecFactory(@Named("SMILE") Provider<ObjectMapper> smileObjectMapperProvider)
```
Add a `@ForSmile` annotation. We avoid using `@Named`.

electrum commented on json/src/main/java/io/airlift/json/SmileCodecFactory.java:
```diff
@@ -0,0 +1,155 @@
+    {
+        ObjectMapper objectMapper = null;
+
+        // Work around for http://code.google.com/p/google-guice/issues/detail?id=627
```
It looks like we don't need this anymore. Can you remove it from the other one in a commit before this one?

electrum commented on json/src/main/java/io/airlift/json/SmileCodecFactory.java:
```diff
@@ -0,0 +1,155 @@
+        this(smileObjectMapperProvider, false);
+    }
+
+    public SmileCodecFactory(Provider<ObjectMapper> smileObjectMapperProvider, boolean prettyPrint)
```
Does Smile actually have pretty print?

electrum commented on json/src/test/java/io/airlift/json/TestSmileCodecFactory.java:
```diff
@@ -0,0 +1,199 @@
+
+    private void validatePersonListSmileCodec(JsonCodec<List<Person>> smileCodec)
+    {
+        ImmutableList<Person> expected = ImmutableList.of(
```
Nit: declare these as `List`, `Map`, etc.

electrum commented on http-client/src/main/java/io/airlift/http/client/FullJsonResponseHandler.java:
```diff
@@ -63,7 +68,7 @@ private FullJsonResponseHandler(JsonCodec<T> jsonCodec)
         byte[] bytes = readResponseBytes(response);
         String contentType = response.getHeader(CONTENT_TYPE);
-        if ((contentType == null) || !MediaType.parse(contentType).is(MEDIA_TYPE_JSON)) {
+        if ((contentType == null) || !SUPPORTED_MEDIA_TYPES.contains(MediaType.parse(contentType))) {
```
This doesn't seem right. The `JsonCodec` in this handler supports either JSON or Smile, but not both, so we shouldn't accept both. Also, the error handling logic and `getJson()` API (used by error handling) assumes JSON. It wouldn't work for Smile which is binary.

electrum reviewed (commented):
While supporting multiple encodings in `JsonCodec` seems attractive, the API methods `fromJson()` and `toJson()` take `String` which is only appropriate for JSON, not Smile (or other binary formats).

I think we need to think more about the APIs and how they are used in applications. Another consideration is that the choice of encoding between client and server should ideally be handled by content negotiation, but we don't have a "JSON HTTP client" abstraction in Airlift, so we can't control the request side.

nezihyigitbasi commented:
> I think we need to think more about the APIs and how they are used in applications.

Agreed. This initial patch is  an easy-way-out kind of patch to add beta support for smile encoding. I also thought about having separate classes like SmileCodec, SmileCodecFactory, etc., which may be a better way to handle it.

> Another consideration is that the choice of encoding between client and server should ideally be handled by content negotiation, but we don't have a "JSON HTTP client" abstraction in Airlift, so we can't control the request side.

Currently in my tests the http clients set the content type/accept headers themselves, and the server is also updated to support both json/smile.

johngmyers commented:
See also `SmileMapper`, `SmileBodyGenerator`, `SmileResponseHandler`, etc. in proofpoint/platform.

nezihyigitbasi commented:
I updated the PR @martint, please review when you have time. I also updated the title to be "WIP" as we need to sort out API details and I need to do more testing with all these latest changes that I did.

I created different classes for smile support and that cleaned up the code quite a bit. But, at the same time it made it a bit difficult to support both json and smile on the client side (say, based on a config flag) as we have separate `SmileCodec` and `JsonCodec` classes now, and the client needs to decide which codec it needs to use up front (e.g., what should be the type of `HttpRemoteTask.taskInfoCodec`?). Maybe we should have a base class/interface for these two codec classes, so the clients can just that type.

johngmyers commented on json/src/main/java/io/airlift/json/SmileCodec.java:
```diff
@@ -0,0 +1,161 @@
+    }
+
+    /**
+     * Coverts the specified smile bytes (UTF-8) into an instance of type T.
```
typo

nezihyigitbasi commented:
@martint I updated the PR with a fixup commit (I can squash if that's what you prefer):
- fix the typo
- add a `SmileBodyGenerator`
- add a `Codec` to make clients more flexible to be able to use both json and smile codecs (say, depending on some config). 

I tested this change with Presto (with a config that enables using smile encoding and with REST endpoints that support both json and smile), so it's no longer a WIP PR.

nezihyigitbasi commented on http-client/src/main/java/io/airlift/http/client/FullSmileResponseHandler.java:
```diff
@@ -0,0 +1,211 @@
+import static io.airlift.http.client.ResponseHandlerUtils.propagate;
+import static java.util.Objects.requireNonNull;
+
+public class FullSmileResponseHandler<T>
```
Modeled after `FullJsonResponseHandler`.

arhimondr replied:
This can be `A___ResponseHandler`, that accepts `Codec`, and uses `Codec#fromBytes` to decode response. So it can be used for both, Json and Smile codec

nezihyigitbasi commented on http-client/src/main/java/io/airlift/http/client/SmileBodyGenerator.java:
```diff
@@ -0,0 +1,32 @@
+import io.airlift.json.SmileCodec;
+
+@Beta
+public class SmileBodyGenerator<T>
```
Modeled after `JsonBodyGenerator`.

nezihyigitbasi commented on json/src/main/java/io/airlift/json/SmileCodec.java:
```diff
@@ -0,0 +1,162 @@
+import static java.util.Objects.requireNonNull;
+
+@Beta
+public class SmileCodec<T>
```
Modeled after `JsonCodec`.

arhimondr replied:
Do we need all of these static factories? I think it should be disencouraged. Usually we should use `Factory`.

@dain ? @electrum ?

nezihyigitbasi commented on json/src/main/java/io/airlift/json/SmileCodecBinder.java:
```diff
@@ -0,0 +1,97 @@
+import static java.util.Objects.requireNonNull;
+
+@Beta
+public class SmileCodecBinder
```
Modeled after `JsonCodecBinder`.

arhimondr replied:
This can be abstracted

nezihyigitbasi commented on json/src/main/java/io/airlift/json/SmileCodecFactory.java:
```diff
@@ -0,0 +1,119 @@
+import static java.util.Objects.requireNonNull;
+
+@Beta
+public class SmileCodecFactory
```
Modeled after `JsonCodecFactory`.

arhimondr replied:
It can also be abstracted

nezihyigitbasi commented on json/src/main/java/io/airlift/json/SmileCodecProvider.java:
```diff
@@ -0,0 +1,70 @@
+
+import static java.util.Objects.requireNonNull;
+
+class SmileCodecProvider
```
Modeled after `JsonCodecProvider`.

nezihyigitbasi reviewed (commented):
Per our offline discussion I commented on the classes that are modeled after the json counterparts.

electrum commented:
I merged the first commit that removes the Guava workaround.

arhimondr commented on json/src/main/java/io/airlift/json/Codec.java:
```diff
@@ -0,0 +1,19 @@
+ */
+package io.airlift.json;
+
+public interface Codec<T>
```
If you had a method for deserialization here as well, (e.g.: ` public T fromJson(byte[] bytes)`  -> ` public T fromBytes(byte[] bytes)` ) you can nicely abstract a lot of stuff.

electrum replied:
Why not have a `T fromBytes(byte[] bytes)`  method as well? This would allow fully generic use of the codec.

arhimondr commented on http-client/src/main/java/io/airlift/http/client/ResponseHandler.java:
```diff
@@ -25,4 +28,27 @@ T handleException(Request request, Exception exception)
     T handle(Request request, Response response)
             throws E;
+
+    interface BaseResponse<T>
```
With `A_ResponseHandler` approach this won't be needed.

electrum replied:
This interface seems rather specific to put in `ResponseHandler`. Maybe a top level `ValueResponse`?

arhimondr commented on http-client/src/main/java/io/airlift/http/client/SmileBodyGenerator.java:
```diff
@@ -0,0 +1,32 @@
+
+    private SmileBodyGenerator(SmileCodec<T> smileCodec, T instance)
+    {
+        super(smileCodec.toBytes(instance));
```
Nice

arhimondr commented on http-client/src/test/java/io/airlift/http/client/TestFullSmileResponseHandler.java:
```diff
@@ -0,0 +1,160 @@
+import static org.testng.Assert.assertTrue;
+import static org.testng.Assert.fail;
+
+public class TestFullSmileResponseHandler
```
This can also be abstracted into an `AbstractTestA__ResponseHandler`. The Codec creation can be abstracted behind factory, and specialized in the `TestResponseHandlerSmile` and `TestResponseHandlerJson`

arhimondr commented on json/src/test/java/io/airlift/json/TestSmileCodecFactory.java:
```diff
@@ -0,0 +1,199 @@
+
+import static org.testng.Assert.assertEquals;
+
+public class TestSmileCodecFactory
```
This can be abstracted

arhimondr commented on json/src/test/java/io/airlift/json/TestSmileSupport.java:
```diff
@@ -0,0 +1,149 @@
+import static io.airlift.json.TestJsonModule.CAR;
+import static org.testng.Assert.assertEquals;
+
+public class TestSmileSupport
```
This can be abstracted

nezihyigitbasi reviewed (commented):
I defined several abstract classes to pull up the common code and simplify the PR. Thanks @arhimondr.
